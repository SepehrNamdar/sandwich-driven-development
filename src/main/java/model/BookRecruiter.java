package model;

import java.time.LocalDate;
import java.util.List;

// Domain Service
public class BookRecruiter {
    public Interview planInterview(
            LocalDate availability, Candidate candidate, List<Recruiter> availableRecruiters, RecruiterRepository recruiters) {
        List<String> candidateSkills = candidate.getSkills();
        Recruiter appropriateRecruiter =
                candidate.findAppropriateRecruiter(availableRecruiters, candidateSkills);
        Recruiter recruiter = recruiters.bookAvailability(appropriateRecruiter, availability);// is a shared state ? Yes
        return new Interview().getInterview(availability, candidate, recruiter);
    }
}
