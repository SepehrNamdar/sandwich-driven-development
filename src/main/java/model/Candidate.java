package model;

import java.util.List;
import java.util.Objects;

// Anemic
public class Candidate {
    private List<String> skills;
    private String name;

    // Sans comportement

    public List<String> getSkills() {
        return skills;
    }

    public void setSkills(List<String> skills) {
        this.skills = skills;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Candidate candidate = (Candidate) o;
        return Objects.equals(skills, candidate.skills) &&
                Objects.equals(name, candidate.name);
    }

    @Override
    public int hashCode() {
        return Objects.hash(skills, name);
    }

    @Override
    public String toString() {
        return "Candidate{" +
                "skills=" + skills +
                ", name='" + name + '\'' +
                '}';
    }
}
