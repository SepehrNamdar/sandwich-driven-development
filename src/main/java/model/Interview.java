package model;

import java.time.LocalDate;
import java.util.List;
import java.util.Objects;

// Aggregate
public class Interview {
    private Candidate candidate;
    private Recruiter recruiter;
    private LocalDate interviewDate;

    public Interview getInterview(LocalDate availability, Candidate candidate, List<Recruiter> availableRecruiters) {
        List<String> candidateSkills = candidate.getSkills();   // Shared state ? NON

        Recruiter appropriateRecruiter = candidate.getAppropriateRecruiter(availableRecruiters, candidateSkills);

        Interview interview = new Interview();
        interview.setCandidate(candidate);
        interview.setRecruiter(appropriateRecruiter);
        interview.setInterviewDate(availability);
        return interview;
    }

    public Candidate getCandidate() {
        return candidate;
    }

    public void setCandidate(Candidate candidate) {
        this.candidate = candidate;
    }

    public Recruiter getRecruiter() {
        return recruiter;
    }

    public void setRecruiter(Recruiter recruiter) {
        this.recruiter = recruiter;
    }

    public LocalDate getInterviewDate() {
        return interviewDate;
    }

    public void setInterviewDate(LocalDate interviewDate) {
        this.interviewDate = interviewDate;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Interview interview = (Interview) o;
        return Objects.equals(candidate, interview.candidate) &&
                Objects.equals(recruiter, interview.recruiter) &&
                Objects.equals(interviewDate, interview.interviewDate);
    }

    @Override
    public int hashCode() {
        return Objects.hash(candidate, recruiter, interviewDate);
    }

    @Override
    public String toString() {
        return "Interview{" +
                "candidate=" + candidate +
                ", recruiter=" + recruiter +
                ", interviewDate=" + interviewDate +
                '}';
    }
}
