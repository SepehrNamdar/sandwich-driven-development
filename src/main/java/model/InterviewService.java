package model;

import java.time.LocalDate;
import java.util.List;

// Domain Service
public class InterviewService {


}
