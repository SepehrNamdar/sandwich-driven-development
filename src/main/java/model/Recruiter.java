package model;

import java.time.LocalDate;
import java.util.List;
import java.util.Objects;

import static java.util.stream.Collectors.toList;

// Rich object
public class Recruiter {
    private String recruiterId;
    private List<String> skills;
    private List<LocalDate> availabilities;
    private String name;

    // behaviour
    public List<Recruiter> getAvailableRecruiters(LocalDate availability, List<Recruiter> allRecruiters) {
        return allRecruiters.stream()
                .filter(recruiter -> recruiter.getAvailabilities().contains(availability))
                .collect(toList());
    }

    public void bookAvailability(LocalDate availability) {
        getAvailabilities().remove(availability);
    }

    boolean canTest(List<String> candidateSkills) {
        return getSkills().containsAll(candidateSkills);
    }

    public List<String> getSkills() {
        return skills;
    }

    public void setSkills(List<String> skills) {
        this.skills = skills;
    }

    public List<LocalDate> getAvailabilities() {
        return availabilities;
    }

    public void setAvailabilities(List<LocalDate> availabilities) {
        this.availabilities = availabilities;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getRecruiterId() {
        return recruiterId;
    }

    public void setRecruiterId(String recruiterId) {
        this.recruiterId = recruiterId;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Recruiter recruiter = (Recruiter) o;
        return Objects.equals(recruiterId, recruiter.recruiterId) &&
                Objects.equals(skills, recruiter.skills) &&
                Objects.equals(availabilities, recruiter.availabilities) &&
                Objects.equals(name, recruiter.name);
    }

    @Override
    public int hashCode() {
        return Objects.hash(recruiterId, skills, availabilities, name);
    }

    @Override
    public String toString() {
        return "Recruiter{" +
                "recruiterId='" + recruiterId + '\'' +
                ", skills=" + skills +
                ", availabilities=" + availabilities +
                ", name='" + name + '\'' +
                '}';
    }
}
