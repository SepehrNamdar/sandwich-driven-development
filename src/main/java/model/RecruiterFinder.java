package model;

import use_case.AnyRecruiterFoundException;
import use_case.RecruiterRepository;

import java.time.LocalDate;
import java.util.List;

// Domain Service
public class RecruiterFinder {
    public Recruiter findAppropriateRecruiter(
            LocalDate availability, Candidate candidate, List<Recruiter> availableRecruiters, RecruiterRepository recruiters) {
        List<String> candidateSkills = candidate.getSkills();

        Recruiter appropriateRecruiter = availableRecruiters.stream()
                .filter(availableRecruiter -> availableRecruiter.canTest(candidateSkills))
                .findFirst()
                .orElseThrow(AnyRecruiterFoundException::new);

        return recruiters.bookAvailability(appropriateRecruiter, availability);
    }
}
