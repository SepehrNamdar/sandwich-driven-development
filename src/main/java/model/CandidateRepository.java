package model;

import model.Candidate;

public interface CandidateRepository {
    Candidate findById(String candidateId);
}
