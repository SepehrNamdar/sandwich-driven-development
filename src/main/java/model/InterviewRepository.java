package model;

import model.Interview;

public interface InterviewRepository {
    void save(Interview interview);
}
