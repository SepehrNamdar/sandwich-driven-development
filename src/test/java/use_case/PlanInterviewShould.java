package use_case;

import model.Interview;
import org.junit.jupiter.api.Test;

import java.time.LocalDate;

import static java.time.LocalDate.of;
import static org.approvaltests.combinations.CombinationApprovals.verifyAllCombinations;

public class PlanInterviewShould {

    @Test
    void find_an_available_recruiter_who_can_test_candidate() {
        String[] candidateIds = {"123", "456", "789", "", null, "BAD-ID"};
        LocalDate[] interviewDates = {
                of(2021, 2, 20),
                of(2021, 2, 21),
                of(2021, 2, 22),
                of(2033, 2, 20),
                of(1789, 2, 20),
                null
        };

        verifyAllCombinations(this::planInterview, candidateIds, interviewDates);
    }

    private Interview planInterview(String candidateId, LocalDate interviewDate) {
        // GIVEN
        PlanInterview interview =
                new PlanInterview(new FakeCandidates(), new FakeRecruiters(), new FakeInterview());

        // WHEN
        return interview.plan(candidateId, interviewDate);
    }

}
