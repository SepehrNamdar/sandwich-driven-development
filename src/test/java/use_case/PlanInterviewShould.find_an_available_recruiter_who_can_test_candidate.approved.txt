[123, 2021-02-20] => Interview{candidate=Candidate{skills=[.Net, Java, PHP, JS], name='Alex'}, recruiter=Recruiter{recruiterId='002', skills=[Java, .Net, PHP, JS], availabilities=[2021-02-22], name='Mary'}, interviewDate=2021-02-20} 
[123, 2021-02-21] => use_case.AnyRecruiterFoundException: null 
[123, 2021-02-22] => Interview{candidate=Candidate{skills=[.Net, Java, PHP, JS], name='Alex'}, recruiter=Recruiter{recruiterId='002', skills=[Java, .Net, PHP, JS], availabilities=[2021-02-20], name='Mary'}, interviewDate=2021-02-22} 
[123, 2033-02-20] => use_case.AnyRecruiterFoundException: null 
[123, 1789-02-20] => use_case.AnyRecruiterFoundException: null 
[123, null] => use_case.AnyRecruiterFoundException: null 
[456, 2021-02-20] => Interview{candidate=Candidate{skills=[JS], name='Bob'}, recruiter=Recruiter{recruiterId='001', skills=[PHP, JS], availabilities=[], name='Emma'}, interviewDate=2021-02-20} 
[456, 2021-02-21] => use_case.AnyRecruiterFoundException: null 
[456, 2021-02-22] => Interview{candidate=Candidate{skills=[JS], name='Bob'}, recruiter=Recruiter{recruiterId='002', skills=[Java, .Net, PHP, JS], availabilities=[2021-02-20], name='Mary'}, interviewDate=2021-02-22} 
[456, 2033-02-20] => use_case.AnyRecruiterFoundException: null 
[456, 1789-02-20] => use_case.AnyRecruiterFoundException: null 
[456, null] => use_case.AnyRecruiterFoundException: null 
[789, 2021-02-20] => use_case.AnyRecruiterFoundException: null 
[789, 2021-02-21] => use_case.AnyRecruiterFoundException: null 
[789, 2021-02-22] => use_case.AnyRecruiterFoundException: null 
[789, 2033-02-20] => use_case.AnyRecruiterFoundException: null 
[789, 1789-02-20] => use_case.AnyRecruiterFoundException: null 
[789, null] => use_case.AnyRecruiterFoundException: null 
[, 2021-02-20] => java.lang.NullPointerException: Cannot invoke "model.Candidate.getSkills()" because "candidate" is null 
[, 2021-02-21] => java.lang.NullPointerException: Cannot invoke "model.Candidate.getSkills()" because "candidate" is null 
[, 2021-02-22] => java.lang.NullPointerException: Cannot invoke "model.Candidate.getSkills()" because "candidate" is null 
[, 2033-02-20] => java.lang.NullPointerException: Cannot invoke "model.Candidate.getSkills()" because "candidate" is null 
[, 1789-02-20] => java.lang.NullPointerException: Cannot invoke "model.Candidate.getSkills()" because "candidate" is null 
[, null] => java.lang.NullPointerException: Cannot invoke "model.Candidate.getSkills()" because "candidate" is null 
[null, 2021-02-20] => java.lang.NullPointerException: Cannot invoke "model.Candidate.getSkills()" because "candidate" is null 
[null, 2021-02-21] => java.lang.NullPointerException: Cannot invoke "model.Candidate.getSkills()" because "candidate" is null 
[null, 2021-02-22] => java.lang.NullPointerException: Cannot invoke "model.Candidate.getSkills()" because "candidate" is null 
[null, 2033-02-20] => java.lang.NullPointerException: Cannot invoke "model.Candidate.getSkills()" because "candidate" is null 
[null, 1789-02-20] => java.lang.NullPointerException: Cannot invoke "model.Candidate.getSkills()" because "candidate" is null 
[null, null] => java.lang.NullPointerException: Cannot invoke "model.Candidate.getSkills()" because "candidate" is null 
[BAD-ID, 2021-02-20] => java.lang.NullPointerException: Cannot invoke "model.Candidate.getSkills()" because "candidate" is null 
[BAD-ID, 2021-02-21] => java.lang.NullPointerException: Cannot invoke "model.Candidate.getSkills()" because "candidate" is null 
[BAD-ID, 2021-02-22] => java.lang.NullPointerException: Cannot invoke "model.Candidate.getSkills()" because "candidate" is null 
[BAD-ID, 2033-02-20] => java.lang.NullPointerException: Cannot invoke "model.Candidate.getSkills()" because "candidate" is null 
[BAD-ID, 1789-02-20] => java.lang.NullPointerException: Cannot invoke "model.Candidate.getSkills()" because "candidate" is null 
[BAD-ID, null] => java.lang.NullPointerException: Cannot invoke "model.Candidate.getSkills()" because "candidate" is null 
